tools/lsp/preview/ui.rs

```diff
@@ -563,34 +571,181 @@ fn map_properties_to_ui(
         property_groups.push(property_group_from(&current_group, current_group_properties));
     }
 
-    Some(ElementInformation {
-        id: element.id.clone().into(),
-        type_name: element.type_name.clone().into(),
-        source_uri,
-        source_version,
-        range: to_ui_range(element.range)?,
+    Some((
+        ElementInformation {
+            id: element.id.clone().into(),
+            type_name: element.type_name.clone().into(),
+            source_uri,
+            source_version,
+            range: to_ui_range(element.range)?,
+        },
+        declarations,
+        Rc::new(VecModel::from(property_groups)).into(),
+    ))
+}
 
-        properties: Rc::new(VecModel::from(property_groups)).into(),
-    })
+enum Op {
+    Insert((usize, usize)),
+    Copy((usize, usize)),
+    PushBack(usize),
+    Remove(usize),
+}
+
+fn is_equal_value(c: &PropertyValue, n: &PropertyValue) -> bool {
+    c.value_bool == n.value_bool
+        && c.is_translatable == n.is_translatable
+        && c.kind == n.kind
+        && c.value_brush == n.value_brush
+        && c.value_float == n.value_float
+        && c.value_int == n.value_int
+        && c.default_selection == n.default_selection
+        && c.value_string == n.value_string
+        && c.tr_context == n.tr_context
+        && c.tr_plural == n.tr_plural
+        && c.tr_plural_expression == n.tr_plural_expression
+        && c.visual_items.row_count() == n.visual_items.row_count()
+        && std::iter::zip(c.visual_items.iter(), n.visual_items.iter()).all(|(c, n)| c == n)
+}
+
+fn is_equal(c: &PropertyInformation, n: &PropertyInformation) -> bool {
+    c.name == n.name && c.type_name == n.type_name && is_equal_value(&c.value, &n.value)
+}
+
+fn is_equal_element(c: &ElementInformation, n: &ElementInformation) -> bool {
+    c.id == n.id
+        && c.type_name == n.type_name
+        && c.source_uri == n.source_uri
+        && c.range.start == n.range.start
+}
+
+pub type PropertyGroupModel = slint::ModelRc<PropertyGroup>;
+
+fn update_properties(
+    current_model: PropertyGroupModel,
+    next_model: PropertyGroupModel,
+) -> PropertyGroupModel {
+    debug_assert_eq!(current_model.row_count(), next_model.row_count());
+
+    for (c, n) in std::iter::zip(current_model.iter(), next_model.iter()) {
+        debug_assert_eq!(c.group_name, n.group_name);
+
+        let cvg = c.properties.as_any().downcast_ref::<VecModel<PropertyInformation>>().unwrap();
+        let nvg = n.properties.as_any().downcast_ref::<VecModel<PropertyInformation>>().unwrap();
+
+        let mut to_do = Vec::new();
+
+        let mut c_it = cvg.iter();
+        let mut n_it = nvg.iter();
+
+        let mut cp = c_it.next();
+        let mut np = n_it.next();
+
+        let mut c_index = 0_usize;
+        let mut n_index = 0_usize;
+
+        loop {
+            match (cp.as_ref(), np.as_ref()) {
```

Comment:
Since each branch only emit one or zero operation, you don't need to_do, i think, you could process the operation dirrectly.